Add changelog for release-1.4.0 (#1543)
astronomer/astro-sdk: python-sdk/docs/CHANGELOG.md

          metadata=Metadata(
              database=os.environ["SNOWFLAKE_DATABASE"],
              schema=os.environ["SNOWFLAKE_SCHEMA"],
          ),
          use_native_support=False,
      ),
  )
  ```
- Add databricks support [#1352](https://github.com/astronomer/astro-sdk/pull/1352), [#1397](https://github.com/astronomer/astro-sdk/pull/1397), [#1452](https://github.com/astronomer/astro-sdk/pull/1452), [#1476](https://github.com/astronomer/astro-sdk/pull/1476), [#1480](https://github.com/astronomer/astro-sdk/pull/1480), [#1555](https://github.com/astronomer/astro-sdk/pull/1555)
  ```python
  aql.load_file(
      input_file=File(
          "s3://astro-sdk/python_sdk/example_dags/data/sample.csv", conn_id="aws_conn"
      ),
      output_table=Table(
          conn_id=DATABRICKS_CONN_ID,
      ),
      load_options=[
          DeltaLoadOptions(
              copy_into_format_options={"header": "true", "inferSchema": "true"},
              copy_into_copy_options={"mergeSchema": "true"},
          )
      ],
  )
  ```
- Add source code facet for openlineage on operators dataframe and transform [#1537](https://github.com/astronomer/astro-sdk/pull/1537)

> - Use load_options param to pass pandasOptions in load_file operator [#1466](https://github.com/astronomer/astro-sdk/pull/1466)

Reviewer: ```suggestion
- Enhance `LoadFileOperator` so that users can send pandas attributes through `PandasCsvLoadOptions` [#1466](https://github.com/astronomer/astro-sdk/pull/1466)
```

          "s3://astro-sdk/python_sdk/example_dags/data/sample.csv", conn_id="aws_conn"
      ),
      output_table=Table(
          conn_id=SNOWFLAKE_CONN_ID,
          metadata=Metadata(
              database=os.environ["SNOWFLAKE_DATABASE"],
              schema=os.environ["SNOWFLAKE_SCHEMA"],
          ),
      ),
      use_native_support=False,
      load_options=PandasCsvLoadOptions(delimiter="$"),
  )
  ```
- Add a snowflake load options in order to pass native API supported param in load_file operator [#1516](https://github.com/astronomer/astro-sdk/pull/1516)
  ```python
  aql.load_file(
      input_file=File(
          "s3://astro-sdk/python_sdk/example_dags/data/sample.csv", conn_id="aws_conn"
      ),
      output_table=Table(
          conn_id=SNOWFLAKE_CONN_ID,
      ),
      load_options=[
          SnowflakeLoadOptions(
              file_options={"SKIP_HEADER": 1, "SKIP_BLANK_LINES": True},
              copy_options={"ON_ERROR": "CONTINUE"},
          )
      ],
  )
  ```
- Expose `get_file_list_func` to users that will return iterable File list from given destination file storage [#1380](https://github.com/astronomer/astro-sdk/pull/1380)

### Improvements

> - Deprecate `export_table_to_file` to `export_to_file` (*`ExportTableToFileOperator` and `export_table_to_file` operator would be removed in astro-python-sdk 1.5.0*) [#1503](https://github.com/astronomer/astro-sdk/pull/1503)

Reviewer: ```suggestion
- Deprecate `export_table_to_file` in favor of `export_to_file` (*`ExportTableToFileOperator` and `export_table_to_file` operator would be removed in astro-python-sdk 1.5.0*) [#1503](https://github.com/astronomer/astro-sdk/pull/1503)
```

> - Fix the dataframe test function name [#1490](https://github.com/astronomer/astro-sdk/pull/1490)

Reviewer: Move this to Misc section

### Bug fixes

> - load_file operator should also check for conn_type of conn_id provided to file object [#1471](https://github.com/astronomer/astro-sdk/issues/1471)

Reviewer 2: ```suggestion
- `load_file` operator checks for `conn_type` and `conn_id` provided to file object [#1471](https://github.com/astronomer/astro-sdk/issues/1471)
```

> - load_file operator should also check for conn_type of conn_id provided to file object [#1471](https://github.com/astronomer/astro-sdk/issues/1471)

Reviewer: ```suggestion
- `LoadFileOperator` should check for conn_type of conn_id provided to file object [#1471](https://github.com/astronomer/astro-sdk/issues/1471)
```

- Generate constraints on releases and pushes (not PRs) [#1472](https://github.com/astronomer/astro-sdk/pull/1472)

### Docs

> - Use the latest name of export op in doc [#1477](https://github.com/astronomer/astro-sdk/pull/1477)

Reviewer: ```suggestion
- Change `export_file` to `export_table_to_file`  in the documentation [#1477](https://github.com/astronomer/astro-sdk/pull/1477)
```

# Changelog

## 1.4.0

### Feature:
- Support for SFTP as file location [#1481](https://github.com/astronomer/astro-sdk/pull/1481)

> - Support for FTP as file location [#1482](https://github.com/astronomer/astro-sdk/pull/1482)

Reviewer 2: Example of this is needed for user

Reviewer 3: FYI: We can also point to documentation

- Add support for Azure Blob Storage [#1275](https://github.com/astronomer/astro-sdk/pull/1275), [#1542](https://github.com/astronomer/astro-sdk/pull/1542)
- Add databricks support [#1352](https://github.com/astronomer/astro-sdk/pull/1352), [#1397](https://github.com/astronomer/astro-sdk/pull/1397), [#1452](https://github.com/astronomer/astro-sdk/pull/1452), [#1476](https://github.com/astronomer/astro-sdk/pull/1476), [#1480](https://github.com/astronomer/astro-sdk/pull/1480), [#1555](https://github.com/astronomer/astro-sdk/pull/1555)

> - Add source code facet for operators [#1537](https://github.com/astronomer/astro-sdk/pull/1537)

Reviewer 2: Which operators should be mentioned here with example

> - Use load_options param to pass pandasOptions in load_file [#1466](https://github.com/astronomer/astro-sdk/pull/1466)

Reviewer 2: Example of this is needed for user

Author: WDYT about https://github.com/astronomer/astro-sdk/pull/1543#discussion_r1067262020

Reviewer 2: https://github.com/astronomer/astro-sdk/releases/tag/1.3.0

> - Add a snowflake load options [#1516](https://github.com/astronomer/astro-sdk/pull/1516)

Reviewer 2: Example of this would be needed for user

Author: WDYT about https://github.com/astronomer/astro-sdk/pull/1543#discussion_r1067262020

Reviewer 2: Added comment above

Reviewer: Add snowflake load options doesnt mean much from user point of view...mention the use of it...something like
"Add snowflake load options in order to....."

> - Expose `get_file_list_func` to users [#1380](https://github.com/astronomer/astro-sdk/pull/1380)

Reviewer 2: This is not clear from user perspective.

### Improvements

> - Refactor snowflake merge function for easier maintenance [#1493](https://github.com/astronomer/astro-sdk/pull/1493)

Reviewer 2: This can go in misc as this is not important from users perspective

# Changelog

## 1.4.0

### Feature:
- Support for SFTP as file location [#1481](https://github.com/astronomer/astro-sdk/pull/1481)
- Support for FTP as file location [#1482](https://github.com/astronomer/astro-sdk/pull/1482)

> - Add support for Azure Blob Storage [#1275](https://github.com/astronomer/astro-sdk/pull/1275)

Reviewer 4: Also: https://github.com/astronomer/astro-sdk/pull/1542

- Add source code facet for operators [#1537](https://github.com/astronomer/astro-sdk/pull/1537)
- Use load_options param to pass pandasOptions in load_file [#1466](https://github.com/astronomer/astro-sdk/pull/1466)
- Add a snowflake load options [#1516](https://github.com/astronomer/astro-sdk/pull/1516)
- Expose get_file_list_func to users [#1380](https://github.com/astronomer/astro-sdk/pull/1380)

### Improvements
- Generate constraints on releases and pushes (not PRs) [#1472](https://github.com/astronomer/astro-sdk/pull/1472)
- Refactor snowflake merge function for easier maintenance [#1493](https://github.com/astronomer/astro-sdk/pull/1493)

> - Deprecate export_table_to_file to export_to_file [#1503](https://github.com/astronomer/astro-sdk/pull/1503)

Reviewer 4: It is probably worth mentioning when we're planning to deprecate `export_table_to_file`

- Fix the dataframe test function name [#1490](https://github.com/astronomer/astro-sdk/pull/1490)
- Decouple unit and integration tests for python-sdk [#1487](https://github.com/astronomer/astro-sdk/pull/1487)

> - Use autospec=True for mocks [#1500](https://github.com/astronomer/astro-sdk/pull/1500)

Reviewer 4: Do Python SDK end-users care about these changes?

Author: same https://github.com/astronomer/astro-sdk/pull/1543#discussion_r1066774377

### Bug fixes
- Loadfile should also check for conn_type of conn_id provided [#1471](https://github.com/astronomer/astro-sdk/issues/1471)

> - Fix pre-commit issues [#1536](https://github.com/astronomer/astro-sdk/pull/1536)

Reviewer 4: Is this change meaningful for end-users?

Author: same as https://github.com/astronomer/astro-sdk/pull/1543#discussion_r1066774377

> - Load/export file from Azure blob storage using non-native path [#1542](https://github.com/astronomer/astro-sdk/pull/1542)

Reviewer 4: This is probably a feature!

Author: I believe  https://github.com/astronomer/astro-sdk/pull/1275 https://github.com/astronomer/astro-sdk/pull/1542 would complete azure storage integration

- CI: Fix circular import error for Airflow 2.2.5 [#1396](https://github.com/astronomer/astro-sdk/pull/1396)

> - Change attrs to attr [#1400](https://github.com/astronomer/astro-sdk/pull/1400)

Author: hmm, does not it make sense to include the changes in the changelog which is meaningful to the contributor also, WDYT?

> - Change attrs to attr [#1400](https://github.com/astronomer/astro-sdk/pull/1400)

Author: give that this is a open-source project

> - Change attrs to attr [#1400](https://github.com/astronomer/astro-sdk/pull/1400)

Reviewer 4: My understanding is that changelogs are usually more user-focused than developer-focused, like Airflow's:
https://airflow.apache.org/docs/apache-airflow/2.5.0/release_notes.html#airflow-2-5-0-2022-12-02

> - Change attrs to attr [#1400](https://github.com/astronomer/astro-sdk/pull/1400)

Author: sure, I'll remove this and other internal PR's

# Changelog

## 1.4.0

### Feature:

> - Support for SFTP as file location [#1481](https://github.com/astronomer/astro-sdk/pull/1481)

Author: here we have PR

> - Support for SFTP as file location [#1481](https://github.com/astronomer/astro-sdk/pull/1481)

Reviewer 2: For astro-sdk if you look at most of them its issues rather than PR: https://raw.githubusercontent.com/astronomer/astro-sdk/main/python-sdk/docs/CHANGELOG.md We should have some consistency there.

- Support for FTP as file location [#1482](https://github.com/astronomer/astro-sdk/pull/1482)

> - Add support for Azure Blob Storage [#1275](https://github.com/astronomer/astro-sdk/pull/1275), [#1542](https://github.com/astronomer/astro-sdk/pull/1542)

Reviewer 3: Native is not done yet, so we should stick to non-native.

> - Add support for Azure Blob Storage [#1275](https://github.com/astronomer/astro-sdk/pull/1275), [#1542](https://github.com/astronomer/astro-sdk/pull/1542)

Reviewer 2: Yes, that's what I mean, we should explicitly specify that for our users

> - Add databricks support [#1352](https://github.com/astronomer/astro-sdk/pull/1352), [#1397](https://github.com/astronomer/astro-sdk/pull/1397), [#1452](https://github.com/astronomer/astro-sdk/pull/1452), [#1476](https://github.com/astronomer/astro-sdk/pull/1476), [#1480](https://github.com/astronomer/astro-sdk/pull/1480), [#1555](https://github.com/astronomer/astro-sdk/pull/1555)

Author: shall I add Significant changes section as airflow does in the minor release?
https://airflow.apache.org/docs/apache-airflow/2.5.0/release_notes.html#airflow-2-5-0-2022-12-02

> - Add databricks support [#1352](https://github.com/astronomer/astro-sdk/pull/1352), [#1397](https://github.com/astronomer/astro-sdk/pull/1397), [#1452](https://github.com/astronomer/astro-sdk/pull/1452), [#1476](https://github.com/astronomer/astro-sdk/pull/1476), [#1480](https://github.com/astronomer/astro-sdk/pull/1480), [#1555](https://github.com/astronomer/astro-sdk/pull/1555)

Reviewer 2: Examples like this would suffice: https://github.com/astronomer/astro-sdk/releases/tag/1.3.0
